### ISSUE TITLE
[FEATURE] Add AESTETIK as a multi-modal backend for identify_spatial_domains

### ISSUE THREAD

**Issue by KalinNonchev** [cafferychen777/ChatSpatial#28]
## Problem

`identify_spatial_domains` currently dispatches `spagcn`, `stagate`, `graphst`, `banksy`, `leiden`, `louvain`. Of those, SpaGCN is the only backend that consumes the H&E image, and it does so through colour-intensity features (`spagcn_s`, `spagcn_b`). There is currently no option that clusters spots from a *learned* morphology representation alongside expression.

## Proposed solution

Add `aestetik` as a method of `identify_spatial_domains`.

AESTETIK is a convolutional autoencoder that learns per-spot representations by jointly using spatial transcriptomics expression, tissue morphology, and spatial context (a neighbourhood grid around each spot). The representations are then clustered, so it fits the exact contract the existing backends already satisfy: domain labels into `adata.obs`, an embedding into `adata.obsm`.

- Paper: "Representation learning for multi-modal spatially resolved transcriptomics data", *Bioinformatics* (2026) — https://doi.org/10.1093/bioinformatics/btag316
- Code: https://github.com/ratschlab/aestetik (MIT)
- PyPI: `aestetik`

The API is sklearn-style and read-only on the input AnnData:

```python
from aestetik import AESTETIK

model = AESTETIK(n_cluster=params.n_domains, morphology_weight=1.5,
                 window_size=3, clustering_method="kmeans")
model.fit(adata)
embedding = model.transform(adata)   # -> adata.obsm["X_aestetik"]
labels    = model.predict(adata)     # -> adata.obs[domain_key]
```

## Implementation sketch

Following `CONTRIBUTING.md` → "Adding a New Analysis Method":

1. `models/data.py` — add `"aestetik"` to the `SpatialDomainParameters.method` `Literal` plus `aestetik_*` fields (`morphology_weight`, `window_size`, `clustering_method`, `latent_dim`, `max_epochs`, `random_state`), matching the existing `banksy_*` / `graphst_*` naming.
2. `tools/spatial_domains.py` — `_identify_domains_aestetik(...)` in the same dispatch chain, gated by `require("aestetik", ctx, feature=...)`, wrapped in `suppress_output()` and the `ThreadPoolExecutor` + `asyncio.wait_for` timeout pattern the other trained backends use. Returns `(domain_labels, "X_aestetik", statistics)`.
3. `utils/dependency_manager.py` — one `DependencyInfo` entry.
4. `pyproject.toml` — an optional extra. Note `aestetik` declares `requires-python >=3.9,<3.14`, so it would need a `; python_version < '3.14'` marker to keep the 3.14 CI leg resolvable. Its runtime deps are `torch`, `torchvision`, `lightning`, `scanpy`, `squidpy`, `anndata`, `scipy`, `numpy`, `pandas`, `matplotlib`, `scikit-learn`, `joblib`, `tqdm` — everything except `torchvision` and `lightning` is already core or already in `[full]`.
5. `docs/advanced/methods-reference.md` row + unit tests with the backend mocked.

## Two things worth deciding before any code

These are the reason I'm opening an issue rather than sending a PR — both are design calls that belong to you.

**1. Where the morphology features come from.** AESTETIK reads them from `adata.obsm["X_pca_morphology"]` (and expression from `adata.obsm["X_pca_transcriptomics"]`); `fit` does not extract them. The package ships `aestetik.utils.utils_morphology.extract_morphology_embeddings(...)`, which crops each spot from a tissue image and embeds it with a caller-supplied torch model, but that needs `pyvips` and a full-resolution image path. ChatSpatial's loader keeps only the downsampled `hires`/`lowres` arrays in `adata.uns["spatial"]`. So either:

  - **(a)** the tool accepts a user-provided `obsm` key and raises a clear `DependencyError`/`DataNotFoundError` with instructions when it's absent — smallest diff, no new image machinery; or
  - **(b)** ChatSpatial grows a small morphology-embedding step that crops from the `hires` image with Pillow (already core) and a torchvision encoder — more useful, but that's a new capability and a new dependency surface.

  I'd start with (a) and leave (b) as a separate decision.

**2. Lattice coordinates.** AESTETIK's neighbourhood grid needs `adata.obs["x_array"]` / `["y_array"]` on an integer lattice: its KDTree lookup keeps a neighbour only at exact distance 0 from `center + (dx, dy)`, so pixel coordinates silently yield an all-imputed grid. For Visium that means `array_row`/`array_col`. `sc.read_visium` provides those in the `filtered_feature_bc_matrix.h5` path, but the manual `_add_spatial_info_to_adata` fallback in `utils/data_loader.py` copies over only `in_tissue` and the pixel coordinates. So the method would either need those columns preserved, or a documented restriction to lattice-based platforms.

## Not proposed (for completeness)

Two related models from the same group predict spatial transcriptomics from H&E rather than analysing measured ST. I don't think either belongs here as things stand, but flagging them so the question is answered rather than left open:

- **DeepSpot** (https://doi.org/10.1101/2025.02.09.25321567 · https://github.com/ratschlab/DeepSpot · MIT) would need a whole-slide-image input path and a new tool category, and its outputs are *predicted*, not measured — every downstream ChatSpatial tool would treat them as measured values. That's a scope decision for you, not a drop-in method.
- **DeepSpot-M** (https://doi.org/10.64898/2026.06.19.26356060 · https://github.com/ratschlab/DeepSpotM) is licensed PolyForm Noncommercial 1.0.0, which is not OSI-approved, and its weights are CC-BY-NC-SA-4.0 and gated behind a Hugging Face access request plus `huggingface-cli login`. Adding it to an MIT-licensed server's extras would quietly attach a non-commercial constraint to ChatSpatial's commercial users, and CI could never exercise it. I'd leave it out.

## Offer

Happy to send a PR for the AESTETIK backend if you're interested, once you've picked an answer for (1) — I'd rather not guess at the morphology-feature contract and hand you a diff you'd have to undo.

**Comment by cafferychen777** [comment]
Hi Kalin — thank you for such a thoughtful and exceptionally well-scoped proposal. We especially appreciate that you identified the morphology-input and lattice-coordinate contracts before starting a PR.

Yes, we'd be happy to accept a PR adding AESTETIK as an optional backend for `identify_spatial_domains`. It is a natural fit for the existing domain-label and embedding contract.

For the first integration, let's choose option (a): ChatSpatial should consume a precomputed morphology embedding from `adata.obsm`. Morphology extraction from tissue images should remain a separate future capability, since it introduces additional decisions around full-resolution image provenance, encoder weights, caching, and system dependencies.

For the initial contract, we'd suggest:

- Add a configurable `aestetik_morphology_key`, defaulting to `X_pca_morphology`.
- Add a configurable `aestetik_transcriptomics_key`, defaulting to ChatSpatial's existing `X_pca`, so the backend composes with `compute_embeddings` instead of silently repeating PCA.
- Raise an actionable `DataNotFoundError` when either representation is missing or invalid.
- In the backend's private working copy, use existing `x_array`/`y_array` when provided; otherwise map preserved Visium lattice coordinates explicitly. Pixel coordinates should not be used as a silent fallback.
- After `fit()`, publish `model.embedding_` and `model.labels_` from that same fitted run, so the stored labels and embedding remain consistent despite the stochastic `transform()` behavior.
- Keep AESTETIK as a bounded optional dependency with the `python_version < "3.14"` marker and clear installation guidance.

We'll handle preservation of `array_row` and `array_col` in ChatSpatial's manual Visium loader separately on our side, so that does not need to be part of your AESTETIK PR.

The timeout, dependency registry, schema, documentation, and mocked-test structure you proposed all look appropriate. We can work through smaller lifecycle details, including temporary Lightning artifacts, during review.

We also agree with your assessment of DeepSpot and DeepSpot-M: both should remain out of scope for this integration.

Thanks again — we'd be very glad to review the PR.

**Comment by cafferychen777** [comment]
The loader-side lattice-coordinate fix has now landed in [`6517187`](https://github.com/cafferychen777/ChatSpatial/commit/6517187), so your AESTETIK PR can assume that manual Visium loading preserves `array_row` and `array_col`.

**Comment by KalinNonchev** [comment]
Thanks — implementing that contract for AESTETIK; the PR stays scoped to `identify_spatial_domains`.

One correction to my earlier framing on the other two.

DeepSpot is [MIT](https://github.com/ratschlab/DeepSpot), pip-installable, no gated weights. And for DeepSpot-M, an optional extra carries no licence constraint for ChatSpatial itself — Apache-2.0 [ToolUniverse](https://github.com/mims-harvard/ToolUniverse/pull/438) and [Biomni](https://github.com/snap-stanford/Biomni/pull/312) both took it on that basis.

Predicted vs measured is a contract rather than a blocker: predictions go to a named slot (`adata.layers["deepspot_predicted"]`, never `X`) with model and provenance in `.uns`.

The capability worth considering: H&E tile in, AnnData with spatial coordinates and predicted expression out — so your existing twenty tools work unchanged on samples that have no assay, which is most archival material. That's a separate input path, not a backend.

Happy to open a separate issue with a concrete design if it's of interest.

**Comment by cafferychen777** [comment]
Thank you for the correction and the additional context on DeepSpot and DeepSpot-M. Your points about the licensing boundary and the distinction between measured and predicted expression are well taken.

The proposed capability — treating H&E-to-expression inference as a separate input path, publishing predictions to a named layer rather than `X`, and retaining model and provenance metadata in `.uns` — is interesting and meaningfully different from an analysis backend. Please do open a separate issue with the concrete design. It would be especially helpful to cover the tile and coordinate input contract, output-layer semantics, model and weight provenance, caching, optional dependency boundaries, runtime expectations, and a testable minimal workflow. We would be glad to evaluate and discuss it on its own merits.

AESTETIK support has now landed in #29, so I am closing this issue as completed. Thank you again for a very constructive contribution and discussion.
